### PR TITLE
feat(Link)!: the "href" property is required

### PR DESCRIPTION
**Pull request by LakeVostok**
UXRFC-1

### CODE REVIEW

File: src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 1]
```diff
@@ -12,13 +12,22 @@ import {BreadcrumbsSeparator} from './BreadcrumbsSeparator';
...
+
+type LinkBreadcrumbsItem = {
+    href: string;
+} & BaseBreadcrumbsItem;
+
+type ButtonBreadcrumbsItem = {
+    href?: undefined;
+} & BaseBreadcrumbsItem;
+
+export type BreadcrumbsItem = LinkBreadcrumbsItem | ButtonBreadcrumbsItem;
```
Comment (amje) [inline src/components/Breadcrumbs/Breadcrumbs.tsx:30]: Isn't the same as before?

File: src/components/Breadcrumbs/BreadcrumbsButton.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+import {block} from '../utils/cn';
+
+const b = block('breadcrumbs');
+
+export function BreadcrumbsButton(props: {
+    title: string;
+    onClick: React.MouseEventHandler<HTMLElement>;
+    children: React.ReactNode;
+}) {
+    return <button {...props} className={b('switcher', {more: true})} />;
```
Comment (amje) [inline src/components/Breadcrumbs/BreadcrumbsButton.tsx:12]: button must have the `type` attribute with the `"button"` value

Comment (amje) [reply src/components/Breadcrumbs/BreadcrumbsButton.tsx:12]: And add please focus styling to the button

File: src/components/Link/README.md [hunk 3]
```diff
@@ -159,32 +156,32 @@ LANDING_BLOCK-->
...
-| className  | HTML `class` attribute                     |                            `string \| undefined`                             |
-| qa         | HTML `data-qa` attribute, used for testing |                            `string \| undefined`                             |
-| ref        | React ref to Link DOM node                 |                `React.ForwardedRef<HTMLElement> \| undefined`                |
+| Name       | Description                                |                           Type                            |  Default   |
+| :--------- | :----------------------------------------- | :-------------------------------------------------------: | :--------: |
+| view       | Link appearance                            |          `"normal" \| "primary" \| "secondary"`           | `"normal"` |
+| visitable  | Display `:visitable` CSS state             |                  `boolean \| undefined`                   |
+| href       | HTML `href` attribute                      |                   `string \| undefined`                   |
```
Comment (amje) [inline src/components/Link/README.md:174]: table formatting broke

File: src/components/Link/__stories__/Link.new.stories.tsx [hunk 4]
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 
 import type {Meta, StoryFn} from '@storybook/react';
 
-import {Link} from '../Link';
+import {Link, type LinkProps} from '../Link';
```
Comment (amje) [inline src/components/Link/__stories__/Link.new.stories.tsx:5]: We're not using this TS syntax

**Review by amje — CHANGES_REQUESTED** [review]

File: src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 1 above)
Comment (LakeVostok) [reply src/components/Breadcrumbs/Breadcrumbs.tsx:30]: It was types product, became types sum. Its more powerful way for union declaration.

File: src/components/Link/__stories__/Link.new.stories.tsx (hunk 4 above)
Comment (LakeVostok) [reply src/components/Link/__stories__/Link.new.stories.tsx:5]: Well, then it should be regulated by the linter. The language provides me with tools, and I use them. In addition, this constuction is used in other components.
I can set up an appropriate rule to ban inline important. But at first glance it looks strange. Why did you decide not to use it?

File: src/components/Breadcrumbs/Breadcrumbs.tsx (hunk 1 above)
Comment (LakeVostok) [reply src/components/Breadcrumbs/Breadcrumbs.tsx:30]: > I mean there is no difference, `href` is optional anyway. What's the point of this change?

Oh, I figure out, why it wasn't obvious to use this kind of approach. Lost it while rebasing, it should have been:
```
type BaseBreadcrumbsItem = {
    text: string;
    items?: BreadcrumbsItem[];
    title?: string;
};

type LinkBreadcrumbsItem = {
    href: string;
    action?: (event: React.MouseEvent<HTMLElement, MouseEvent> | KeyboardEvent) => void;
} & BaseBreadcrumbsItem;

type ButtonBreadcrumbsItem = {
    href?: undefined;
    action: (event: React.MouseEvent<HTMLElement, MouseEvent> | KeyboardEvent) => void;
} & BaseBreadcrumbsItem;
```
So, we describe 2 categories of breadcrumb items: the one is like regular anchor(and should have the href prop), the two is like button, that should have some action(action prop).
So, if user want to create Breadcrumb, that is not like regular link, he will be forced(by typescript) to specify the action prop.
And if the developer(of uikit) needs to add a new behavior for only one of the types of breadcrumb item(e.g. the target attribute for anchor), then he modifies only LinkBreadcrumbsItem and any time in the code he will know, that he works with certain type of item(due to the href prop check).

File: src/components/Link/__stories__/Link.new.stories.tsx (hunk 4 above)
Comment (amje) [reply src/components/Link/__stories__/Link.new.stories.tsx:5]: I remember the main reason was that it requires to use TS 4.5+

**Review by amje — APPROVED** [review]

**Pull request merged**